### ISSUE TITLE
[Q] High CPU usage

### ISSUE THREAD

**Issue by dweidenfeld** [RustCrypto/password-hashing#15]
Hi there,

first of all: thank you for the nice crate!

But there is one thing that I don't get. I've used an scrypt implementation in Java with the same provided settings as you have mentioned in the documentation (15, 8, 1).
The Java implementation is using a lot of memory and some CPU resources. This is what I expected with the given settings.
When I use the rust version it will consume a lot of CPU, but almost no RAM.

I've setup a load test that will create millions of scrypt hashes and the application is just consuming around 60-80m RAM, but ~1600MHz.
Just for the comparison the Java version with almost the same settings consumes 1500m RAM and ~800MHz.

I am using the `simple_crypt()` function.

Thank you for any kind of response/help :)

**Comment by newpavlov** [comment]
Greetings!

With the given parameters scrypt [allocates](https://github.com/RustCrypto/password-hashing/blob/master/scrypt/src/lib.rs#L97) three arrays: one with size of 32 MiB and two with size of 1KiB. Are you sure that Java implementation uses the same parameters? I would say that high CPU usage is normal as we fully utilize CPU core resources. (though I think we can improve performance further) Can you compare performance of Java and Rust implementations with the same parameters?

**Comment by dweidenfeld** [comment]
I think I'll have to correct my initial statement. 

I've now stripped the setup down to the minimum and added some benchmark tests. The memory usage is comparable. 
The rust version also used actix-web for serving the results, maybe this could lead to the high memory usage.

But it is still interesting, that the Java version seeems to be a bit faster than the rust version.

Java version:
```java
    public static String getHash() {
        return SCryptUtil.scrypt("my password", (int) Math.pow(2, 15), 8, 1);
    }
```
```
Result "com.github.dweidenfeld.scrypt_test.Crypter.getHash":
  0.097 ±(99.9%) 0.002 s/op [Average]
  (min, avg, max) = (0.095, 0.097, 0.101), stdev = 0.002
  CI (99.9%): [0.095, 0.098] (assumes normal distribution)


# Run complete. Total time: 00:01:26

Benchmark        Mode  Cnt  Score   Error  Units
Crypter.getHash  avgt   20  0.097 ± 0.002   s/op
```

Rust version:
```rust
fn get_hash() -> String {
    let params = ScryptParams::new(15, 8, 1).expect("scrypt params");

    scrypt_simple("my password", &params).expect("generated hash")
}
```
```
cargo bench
    Finished release [optimized] target(s) in 0.24s
     Running target/release/deps/rust_scrypt_test-b0c80187cfce5885

running 1 test
test tests::bench_scrypt ... bench: 113,158,913 ns/iter (+/- 11,462,209)

test result: ok. 0 passed; 0 failed; 0 ignored; 1 measured; 0 filtered out
```

I'll replace actix with hyper or another web server framework and report my results here afterwards ;)

**Comment by newpavlov** [comment]
We didn't put much effort into `scrypt` optimizations yet, so I am not surprised that established Java implementation can be a bit faster.

More than 70% of time is spent in `salsa20_8` which I believe can be significantly accelerated using explicit SIMD. Unfortunately looks like compiler autovectorization is unable to handle such code. I'll try to work on it eventually.

**Comment by dweidenfeld** [comment]
I think we should close this issue by now, but thank you for the fast response and nice feedback ;)

**Comment by dweidenfeld** [comment]
Just as an update:

I switched from actix-web to hyper and saw that the memory usage decreased drastically. Unfortunately hyper was not as good at handling concurrent requests, so I switched back to actix-web and tuned the settings quite a bit.
Now the memory usage is still kind of high, but I am able to serve thousands of requests concurrently with this scrypt lib.

Thank you again for the support ;)
